core/cafe_service.go

```diff
@@ -147,94 +175,122 @@ func (h *CafeService) FindContact(query *ContactInfoQuery, cafe peer.ID) ([]*pb.
 			Limit:        int32(query.Limit),
 			Wait:         int32(query.Wait),
 		}, nil, false)
-	})
-	if err != nil {
-		return nil, err
 	}
 
-	res := new(pb.CafeContactQueryResult)
-	if err := ptypes.UnmarshalAny(renv.Message.Payload, res); err != nil {
-		return nil, err
+	session := h.datastore.CafeSessions().Get(cafe.Pretty())
+	if session == nil {
+		return errors.New(fmt.Sprintf("could not find session for cafe %s", cafe.Pretty()))
+	}
+
+	env, err := envFactory(session)
+	if err != nil {
+		return err
 	}
 
-	return res.Contacts, nil
+	renvCh, errCh, cancel := h.service.SendHTTPStreamRequest(getCafeHTTPAddr(session), env)
+	for {
+		select {
+		case <-cancelCh:
+			if cancel != nil {
+				fn := *cancel
+				fn()
```

Comment:
A little weird, but only way I could properly cancel reading a stream on the client side (the decoder blocks while waiting for new values)